### PR TITLE
review destroys

### PR DESCRIPTION
**Pull request by mrlevitas**

### CODE REVIEW

File: app/controllers/parks_controller.rb [hunk 1]
```diff
@@ -43,12 +43,14 @@ def update
...
-      flash[:notice] = "You failed to delete park!"
-      render :show
-    else
       flash[:notice] = "You have deleted park successfully!"
       redirect_to root_path
+    else
+      flash[:notice] = "You failed to delete park!"
+      render :show
     end
   end
```
Comment (kaylarr) [inline app/controllers/parks_controller.rb]: We now have `@park = Park.find(params[:id])` in four places in this controller.. Time to abstract it out:

```
class ParksController < ApplicationController
  ...

  def show
    park
    @review = Review.new
  end

  def edit
    park
  end

  def update
    if park.update(park_params)
      flash[:notice] = "You have successfully updated your question!"
      redirect_to park
    else
      flash.now[:notice] = "Invalid input."
      render 'edit'
    end
  end

  def destroy
    if park.destroy
      flash[:notice] = "You have deleted park successfully!"
      redirect_to root_path
    else
      flash[:notice] = "You failed to delete park!"
      render :show
    end
  end

  private

  def park
    @park ||= Park.find(params[:id])
  end

  def park_params
    ...
  end
end
```

Notice with this strategy that the first line in both `update` and `destroy` takes care of (1) finding the Park object by calling `park` method, (2) assigning it to an instance variable accessed by calling `park` method, (3) performing an action on it, and (4) checking to see if the action performed properly.

That's the advantage of using `update` - it calls `save` for us, so we don't need to explicitly save it afterward, and it also returns true or false depending on whether it saved or not (ie whether the operation successfully performed). Same with `destroy` - it returns true if it executed successfully.

File: app/controllers/reviews_controller.rb [hunk 2]
```diff
@@ -34,6 +34,20 @@ def update
...
+    if @review.destroyed?
+      flash[:notice] = "You have deleted review successfully!"
+      redirect_to park_path(@park)
+    else
+      flash[:notice] = "You failed to delete review!"
+      render "parks/show"
+    end
+  end
+
   private
```
Comment (kaylarr) [inline app/controllers/reviews_controller.rb]: Same note and strategy as above, regarding abstracting out `@review = Review.find(params[:id])`

File: spec/features/reviews/04_destroy_review_spec.rb [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+# I want to delete a review
+# So that I can remove it from a park's show page
+
+feature "review creator can edit review information" do
+
+  scenario "Review creator signs in and navigates to edit review page" do
+    review = create(:review)
+    park = review.park
+    user = review.user
+
```
Comment (kaylarr) [inline spec/features/reviews/04_destroy_review_spec.rb]: You're only using these here in one scenario, so not using `let` statements is fine!

File: app/controllers/parks_controller.rb [hunk 4]
```diff
@@ -6,7 +6,7 @@ def index
...
 
   def show
-    @park = Park.find(params[:id])
+    park
```
Comment (davidrf) [inline app/controllers/parks_controller.rb]: Remove this, you should use a callback to run this for the appropriate actions `before_action :park, only: [:show,  :edit, :update, :destroy]`

File: app/controllers/parks_controller.rb [hunk 5]
```diff
@@ -26,13 +26,11 @@ def create
   end
 
   def edit
-    @park = Park.find(params[:id])
+    park
```
Comment (davidrf) [inline app/controllers/parks_controller.rb]: Remove this, you should use a callback to run this for the appropriate actions

File: app/controllers/parks_controller.rb [hunk 6]
```diff
@@ -26,13 +26,11 @@ def create
...
   end
 
   def update
-    @park = Park.find(params[:id])
-    @park.update(park_params)
-    if @park.save
+    if park.update(park_params)
```
Comment (davidrf) [inline app/controllers/parks_controller.rb]: change this to `if @park.update(park_params)`

File: app/controllers/parks_controller.rb [hunk 7]
```diff
@@ -42,18 +40,21 @@ def update
   end
 
   def destroy
-    @park = Park.find(params[:id])
-    if @park.destroyed?
-      flash[:notice] = "You failed to delete park!"
-      render :show
-    else
+    if park.destroy
```
Comment (davidrf) [inline app/controllers/parks_controller.rb]: change this to `if @park.destroy`

File: app/controllers/reviews_controller.rb [hunk 8]
```diff
@@ -34,6 +32,19 @@ def update
     end
   end
 
+  def destroy
+    review
+    park
```
Comment (davidrf) [inline app/controllers/reviews_controller.rb]: Remove both `review` and `park` , you should use a callback to run this for the appropriate actions

File: app/controllers/reviews_controller.rb [hunk 9]
```diff
@@ -23,9 +23,7 @@ def update
     review
     park
```
Comment (davidrf) [inline app/controllers/reviews_controller.rb]: Remove both `review` and `park`, you should use a callback to run this for the appropriate actions

**Pull request merged**